chainer/training/extensions/parameter_statistics.py

```diff
@@ -0,0 +1,146 @@
+import numpy
+import six
+
+from chainer import reporter
+from chainer import training
+from chainer.training import extension
+
+
+class ParameterStatistics(extension.Extension):
+    """Trainer extension to report parameter statistics.
+
+    Statistics are collected and reported for a given :class:`~chainer.Link`
+    or an iterable of :class:`~chainer.Link`s. If a link contains child links,
+    the statistics are reported separately for each child.
+
+    Any function that takes a one-dimensional :class:`numpy.ndarray` or a
+    :class:`cupy.ndarray` and outputs a single or multiple real numbers can be
+    registered to handle the collection of statistics, e.g.
+    :meth:`numpy.ndarray.mean`.
+
+    The keys of reported statistics follow the convention of link name
+    followed by parameter name, attribute name and function name, e.g.
+    ``VGG16Layers/conv1_1/W/data/mean``. They are prepended with an optional
+    prefix and appended with integer indices if the statistics generating
+    function return multiple values.
+
+    Args:
+        links (~chainer.Link or iterable of ~chainer.Link): Link(s) containing
+            the parameters to observe. The link is expected to have a ``name``
+            attribute which is used as a part of the report key.
+        statistics (dict): Dictionary with function name to function mappings.
+            The name is a string and is used as a part of the report key. The
+            function is responsible for generating the statistics.
+        report_params (bool): If ``True``, report statistics for parameter
+            values such as weights and biases.
+        report_grads (bool): If ``True``, report statistics for parameter
+            gradients.
+        prefix (str): Optional prefix to prepend to the report keys.
+        trigger: Trigger that decides when to aggregate the results and report
+            the values.
+    """
+    default_name = 'parameter_statistics'
+    priority = extension.PRIORITY_WRITER
+
+    # Exclude the first '/' since param_name starts with one
+    report_key_template = '{link_name}{param_name}/{attr_name}/{function_name}'
+
+    default_statistics = {
+        'mean': numpy.mean,
+        'std': numpy.std,
+        'min': numpy.min,
+        'max': numpy.max,
+        'zeros': lambda x: numpy.count_nonzero(x == 0),
+        'percentile': lambda x: numpy.percentile(x, (0.13, 2.28, 15.87,
+                                                     50, 84.13, 97.72,
+                                                     99.87))
+    }
+
+    def __init__(self, links, statistics=default_statistics,
+                 report_params=True, report_grads=True, prefix=None,
+                 trigger=(1, 'epoch')):
+
+        if not isinstance(links, (list, tuple)):
+            links = links,
+        self._links = links
+
+        if statistics is None:
```

Comment:
That might be cleaner since the current code relies on running the link on the CPU. We could rename `default_statistics` to something like `standard_statistics` to still keep it accessible.